sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -90,6 +90,132 @@ case class MapKeys(child: Expression)
+    val fields = arrayAT.zipWithIndex.foldRight(List[StructField]()) {
+      (item, list) => {
+        val (arr, idx) = item
+        StructField(s"_$idx", arr.elementType, arr.containsNull) :: list
+      }
+    }
+    StructType(fields)
+  }
+
+  override def dataType: DataType = ArrayType(mountSchema())
+
+  override def prettyName: String = "zip"
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val genericArrayData = classOf[GenericArrayData].getName
+    val genericInternalRow = classOf[GenericInternalRow].getName
+
+    val evals = children.map(_.genCode(ctx))
+    val numArrs = evals.length
+
+    val arrCardinality = ctx.freshName("arrCardinality")
+    val arrVals = ctx.freshName("arrVals")
+
+    val arrayTypes = children.map(_.dataType.asInstanceOf[ArrayType].elementType)
+
+    val inputs = evals.zipWithIndex.map { case (eval, index) =>
+      s"""
+        |${eval.code}
+        |if (!${eval.isNull}) {
+        |  $arrVals[$index] = ${eval.value};
+        |  $arrCardinality[$index] = ${eval.value}.numElements();
+        |} else {
+        |  $arrVals[$index] = null;
+        |  $arrCardinality[$index] = 0;
+        |}
+      """.stripMargin
+    }.mkString("\n")
+
+    val myobject = ctx.freshName("myobject")
+    val biggestCardinality = ctx.freshName("biggestCardinality")
+    val j = ctx.freshName("j")
+    val i = ctx.freshName("i")
+    val args = ctx.freshName("args")
+
+    val fillValue = evals.zipWithIndex.map { case (eval, index) =>
+      s"""
+      |if ($j == ${index}) {
+      |  $myobject[$j] = ${CodeGenerator.getValue(s"$arrVals[$j]", arrayTypes(index), i)};
+      |}
+      """.stripMargin
+    }.mkString("\n")
+
+    ev.copy(s"""
+      |ArrayData[] $arrVals = new ArrayData[$numArrs];
+      |int[] $arrCardinality = new int[$numArrs];
+      |$inputs
+      |int $biggestCardinality = 0;
+      |for (int $i = 0; $i < $numArrs; $i ++) {
+      |  $biggestCardinality = Math.max($biggestCardinality, $arrCardinality[$i]);
+      |}
+      |Object[] $args = new Object[$biggestCardinality];
+      |for (int $i = 0; $i < $biggestCardinality; $i ++) {
+      |  Object[] $myobject = new Object[$numArrs];
+      |  for (int $j = 0; $j < $numArrs; $j ++) {
+      |    if ($arrVals[$j] != null && $arrCardinality[$j] > $i) {
+      |      $fillValue;
+      |    } else {
+      |      $myobject[$j] = null;
+      |    }
+      |  }
+      |  $args[$i] = new $genericInternalRow($myobject);
+      |}
+      |boolean ${ev.isNull} = false;
+      |$genericArrayData ${ev.value} = new $genericArrayData($args);
+    """.stripMargin)
+  }
+
+  override def nullable: Boolean = children.forall(_.nullable)
+
+  override def eval(input: InternalRow): Any = {
+    val inputArrays = children.map(_.eval(input).asInstanceOf[ArrayData])
+    val arrayTypes = children.map(_.dataType.asInstanceOf[ArrayType].elementType)
+    val numberOfArrays = children.length
```

Comment:
this is fixed for each input. Can we compute it as a lazy val in in order to avoid recomputation?